### ISSUE TITLE
Item variants need to work for both stock management and GAPS

### ISSUE THREAD

**Issue by adamdewey** [msupply-foundation/open-msupply#5432]
## What went wrong? 😲

Open mSupply needs to be able to handle item variants holistically so that the system can be used for both stock management and GAPS at the same time. 

It currently doesn't do that. 

There are currently major issues in terms of data being set at the wrong level or illogical duplication / messiness, depending on how you attempt to define item variants. 

### Starting point: 

This is a sample list of a generic item and its variants from the WHO database. 

<img width="1132" alt="image" src="https://github.com/user-attachments/assets/64c31655-b020-4628-b6d4-aa2845725ad8">

You can see several things here: 

- The item variants can have the same commercial name
- The unit is set at the item variant level, not the item level
- The number of doses is set at the item variant level, not the item level

#### Attempting to use system with `item` as the generic item

If we create a single generic item for the sample vaccine (Rotavirus) above then we end up with a hard to use system for stock management. 

1. The item variant is not very visible and it's hard to know what you're dealing with during stock processes
2. Additionally the unit, strength and other properties are incorrectly set at the generic item level rather than the item variant level
3. It's not possible to enter variants with the same commerical name
4. There are # of doses set on the generic item **and** the item variant and they can differ 

#### Attempting to use system with `item` as the item variant

Another approach was to enter every item variant as an item, but this results in having to add variants to the item variant in order to make GAPS work..

#### `item` as the generic item

![image](https://github.com/user-attachments/assets/42b7ff07-d612-43a7-a9fa-97f4a270e274)

Issues with properties being on the generic item, not the item variant: 

![image](https://github.com/user-attachments/assets/11f62092-e178-42fb-a090-eab140db9f36)

Issues with properties missing from item variant: 

![image](https://github.com/user-attachments/assets/bc01cd7c-f922-4cbb-8e74-00a172f767df)

Issues with not being able to see details about item variants during stock management processes: 

<img width="1508" alt="image" src="https://github.com/user-attachments/assets/9bd3e04c-9c12-4f31-b3df-632f37529f90">

<img width="1511" alt="image" src="https://github.com/user-attachments/assets/b492ec12-b055-4038-841a-7d3a21692587">

Issues with incorrect unit being used: 

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/1ae8c4da-54b3-4645-93e8-3d6fec812990">

<img width="1503" alt="image" src="https://github.com/user-attachments/assets/64cd20b1-5574-44da-8307-6f2fd0c9714a">

#### `item` as the item variant

Tried creating an item for every item variant as an alternative approach

![image](https://github.com/user-attachments/assets/27b32daf-15d9-4e85-80df-57462cfed6b2)

But that means that you would have to create an item variant for each item variant so you could capture packaging details for GAPS...

![image](https://github.com/user-attachments/assets/cc9664b4-6fdd-40e5-a2a2-a27893040196)

![image](https://github.com/user-attachments/assets/43899647-8faa-4a43-b18e-25b325ae6668)

## Your environment 🌱

- Open mSupply Version: 2.4.00-RC1
- Legacy mSupply Central Server Version: 7-18-16

**Comment by craigdrown** [comment]
Be careful that we don't double up on the meaning of units
Currently in mSupply for the above example the units would be "doses" - I think you're referring to "presentation" - which should be a separate field on the variant I think?
We usually say the units field answers the question "when I talk about 1 (pack) of this thing, what am I talking about?" 
- e.g.  tablets or a bottle of 100, 
- e.g. grams of a cream or a 20g tube of a cream.

**Comment by adamdewey** [comment]
@craigdrown - I had a big discussion with @louisaw123 about this for vaccines. 

Originally I was also wanting to use "doses" for units, but Louisa talked me around that it should actually be "vials" or "ampoules" etc and that doses should be 'sub-units' 

I spoke with @jagoje about how we could use the item variants to work for countries and we looked at a couple of client vaccine lists.

It seems that the common thing for vaccines is to group by 'presentation + # of doses' and the item variants would just be different manufacturers for each of those 'presentation + # of doses' combinations 

(I'm not sure what terminology to use for the combination of 'presentation + # of doses' as it seems to be unique for vaccines - but WHO distinguishes between 'presentation' (vial, ampoule etc) and '# of doses' (1, 5, 10 etc)

**Comment by louisaw123** [comment]
Oh wait oops ... i just got what you mean now, I forgot how mSupply describes units lol. Hopefully I didn't just confuse you, otherwise just ignore the terminology I've used to describe it all and just take the context if that is helpful

**Comment by adamdewey** [comment]
Thanks Louisa - yeah so in those examples I think we can treat the ml and vials as equivalent, you're just losing a little bit of info on exactly how big the vials are.

But otherwise that aligns with the approach we're suggesting with having 1 item per 'presentation + dose'

**Comment by mark-prins** [comment]
To summarise requested changes, please edit if I have this incorrect:

* Remove `dose` from the item variant
* Add more detail to the item variant selector used in inbound shipments
* Add item variant selector to the stocktake line edit ( TBC )
* Add item variant selector to inventory adjustments ( TBC )

**Comment by lache-melvin** [comment]
> * Add item variant selector to inventory adjustments

Can I understand this one? If I'm increasing/decreasing the quantity of stock on hand for a particular stock line... why would I need to choose the item variant?

**Comment by adamdewey** [comment]
> Can I understand this one? If I'm increasing/decreasing the quantity of stock on hand for a particular stock line... why would I need to choose the item variant?

Good point, I was thinking of the OG inv adjustments where you could introduce new batches

**Comment by jmbrunskill** [comment]
> Can I understand this one? If I'm increasing/decreasing the quantity of stock on hand for a particular stock line... why would I need to choose the item variant?

Arguably fixing the chosen item_variant would be an inventory adjustment of sorts? Feels like a nice way to fix it rather than having to do a stock take? That said, I think as long as there is a least method to introduce the variant for an existing stockline we're fine.

**Comment by adamdewey** [comment]
> Arguably fixing the chosen item_variant would be an inventory adjustment of sorts? Feels like a nice way to fix it rather than having to do a stock take? That said, I think as long as there is a least method to introduce the variant for an existing stockline we're fine.

Looks like you can just switch the variant on the stockline at the moment without doing an inventory adjustment: 

![image](https://github.com/user-attachments/assets/161c0dff-965d-4187-8954-1b019a14f75f)

Whether that's desirable or not is a different question 🤔

**Comment by mark-prins** [comment]
> Can I understand this one? If I'm increasing/decreasing the quantity of stock on hand for a particular stock line... why would I need to choose the item variant?

this is why I left it out of stocktakes in the requirements.
The reason for bringing it in, is that we can add batches during stocktakes too. In particular, when deploying it is common practice to begin with a stocktake in order to introduce all of your stock into the system. In which case - helpful to have variant selection on stocktake - you don't want to force people to use inbound shipments to bring in all stock, especially stock that's sitting in the facility already

**Comment by adamdewey** [comment]
> this is why I left it out of stocktakes in the requirements.
The reason for bringing it in, is that we can add batches during stocktakes too. In particular, when deploying it is common practice to begin with a stocktake in order to introduce all of your stock into the system. In which case - helpful to have variant selection on stocktake - you don't want to force people to use inbound shipments to bring in all stock, especially stock that's sitting in the facility already

Ah, but stocktakes differ to inventory adjustments right?

I think Laché is meaning this functionality: 

![image](https://github.com/user-attachments/assets/b9565306-312b-43cf-9dd4-a30e4928acd9)

Which only allows the user to increase / decrease existing stocklines, not create new stocklines

**Comment by mark-prins** [comment]
yes, that one and adjusting an existing stock line on a stocktake I'm not bothered about.
Even this one:

<img width="1542" alt="Screenshot 2024-11-21 at 5 23 08 PM" src="https://github.com/user-attachments/assets/126ef11e-cbc3-4d90-b2fa-18330e7dcc48">

which is the 'add a new stock line' version of the 'Adjust' button : I'm less concerned about that.

It's the adding of stock lines through a stocktake which would be the biggest issue

**Comment by jmbrunskill** [comment]
Moving to backlog as high priority items are now done.

**Comment by adamdewey** [comment]
> Just a couple of points - I haven't understood every comment, sorry.
i. important that variants aren't foisted on users that don't want to think about them- in OG we'd have a pref to show/hide if we were doing it- or have a default variant of "n/a" that was always selected.
ii. has to work smoothly with non-cold chain stock
iii. I do think variants need to be able to be stocktaken (sp?) and have their own audit trail (maybe not right now- but certainly before they're widely used). That would mean that you need a transaction for all variant changes (e.g. if changing from one variant to another for a stock line you'd need to decrease one and increase the other with an invoice_line record)

Think i and ii are ok. 

iii isn't in place at the moment - you can switch between variants without any audit trail I think

**Comment by jmbrunskill** [comment]
> iii isn't in place at the moment - you can switch between variants without any audit trail I think

Correct, if updated by a stock take we have a record of the changes, but if we just change the stock line we don't, there will be an audit log entry though (i think?) not sure that it's capturing what changed though. This is the same behaviour as how we handle batch numbers. I'm not sure where this might matter, but feels like batch numbers should also get an audit trail if we go down this path?

**Comment by jmbrunskill** [comment]
Removed from milestone as `Must Haves` are done. Keeping issue open as an Epic for now.

**Comment by andreievg** [comment]
Triage, reducing priority as must haves have been done

**Comment by andreievg** [comment]
Needs refinement removed, a new issue carry over can be created for outstanding things and that can be refined if needed

**Comment by adamdewey** [comment]
The solution has been to consider `item` as neither generic nor variant but something in-between. 

E.g. see https://docs.msupply.foundation/docs/introduction/terminology/#items-products-skus

<img width="805" height="547" alt="Image" src="https://github.com/user-attachments/assets/ec71b7b9-8091-45f4-ab4a-9e9c6576b8e0" />

<img width="846" height="250" alt="Image" src="https://github.com/user-attachments/assets/65c4b1a4-ad15-4dc3-85ba-c8377356e7fc" />

So this can be closed for now
